scripts/needs_ci_run

```diff
@@ -1,41 +1,126 @@
 #!/usr/bin/env python3
 import fnmatch
-import os.path
 import sys
 import subprocess
+import requests
 
-CI_IGNORE_FILE='.circleci/ciignore'
-# Looks for files changed in last commit only. This has the drawback
-# of skipping CI if a push's last commit was ciignore-able, even though
-# other commits might need CI
-#
-# TODO: Get all files changed in the last "push". That ensures CI is skipped
-# if ALL commits in a push don't need CI
-GIT_CMD=['git', 'diff', 'HEAD..HEAD~1', '--name-only']
 
-if not os.path.isfile(CI_IGNORE_FILE):
-  # no CI ignore rules configured, run CI
-  sys.exit(0)
+from os import path, environ
+
+
+# jobs defined in .circleci/config.yml
+JOBS = ['build-docker-images',
+        'lint-build-debug',
+        'build-release',
+        'build-osx']
+
+CIRCLE_JOB_STATUS_SUCCESS = 'success'
+
+'''
+Returns a tuple of status & gitsha of the PR that triggered the build.
+status is True if all jobs of the last build were in 'success' state
+otherwise False
+
+Each push to a branch triggers a new CI run, and any previous run is cancelled.
+A CI run starts 4 jobs (see JOBS above) and each job has a unique "workflow ID"
+attached to it. This workflow ID changes for every CI run. In order to get the
+status of the last run, we do the following:
+- Get status of last 8 jobs from the PR branch - 4 jobs will be from the
+  current run and 4 from the previous run
+- Ignore jobs of the current workflow (using workflow ID)
+- The last run is successful if ALL jobs had a status of 'success'
+- Store the gitsha of the last successful run. This is later used to fetch the
+  list of files changed since last CI run
+'''
+def last_build_successful():
+    branch = environ.get('CIRCLE_BRANCH')
+    current_workflow_id = environ.get('CIRCLE_WORKFLOW_ID')
+    token = environ.get('CIRCLE_API_TOKEN')
+
+    if not branch:
+        print('CIRCLE_BRANCH env not set! Unable to do CI skip check')
+        sys.exit(0)
+    if not current_workflow_id:
+        print('CIRCLE_WORKFLOW_ID env not set! Unable to do CI skip check')
+        sys.exit(0)
+    if not token:
+        print('CIRCLE_API_TOKEN env not set! Unable to do CI skip check')
+        sys.exit(0)
+
+    # Get twice the number of jobs configured. There would be len(JOBS) from this
+    # run, while the rest will be from the last run
+    cci_api_url = \
+        f'https://circleci.com/api/v1.1/project/github/valhalla/valhalla/tree/{branch}?circle-token={token}&shallow=true&limit={len(JOBS)*2}'
```

Comment:
[circle API doc](https://circleci.com/docs/api/#recent-builds-for-a-single-project)